src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp

```diff
@@ -58,27 +52,32 @@ void ReplicatedMergeTreeRestartingThread::run()
     if (need_stop)
         return;
 
-    size_t reschedule_period_ms = check_period_ms;
+    /// In case of any exceptions we want to rerun the this task as fast as possible but we also don't want to keep retrying immediately
+    /// in a close loop (as fast as tasks can be processed), so we'll retry in between 100 and 1000 ms
+    const size_t immediately_ms = static_cast<size_t>(std::min(1000u, 100 * (consecutive_check_failures + 1)));
 
     try
     {
         bool replica_is_active = runImpl();
-        if (!replica_is_active)
-            reschedule_period_ms = retry_period_ms;
-    }
-    catch (const Exception & e)
-    {
-        /// We couldn't activate table let's set it into readonly mode
-        partialShutdown();
-        tryLogCurrentException(log, __PRETTY_FUNCTION__);
-
-        if (e.code() == ErrorCodes::REPLICA_STATUS_CHANGED)
-            reschedule_period_ms = 0;
+        if (replica_is_active)
+        {
+            consecutive_check_failures = 0;
+            task->scheduleAfter(check_period_ms);
+        }
+        else
+        {
+            consecutive_check_failures++;
+            task->scheduleAfter(immediately_ms);
+        }
     }
     catch (...)
     {
+        task->scheduleAfter(immediately_ms);
```

Comment:
It will retry each 100-1000ms even if current replica is completely partitioned from a cluster. Should we increase this value a bit on each failure until it reaches 10s (the original value of `retry_period_ms`)?

Reply:
Sure. I went for 1s because it was reduced to it before, but it should be better to have some exponential backoff (or some other simpler backoff) for up to 10 seconds.